### PR TITLE
[336] Afficher les groupes et les utilisateurs qui existent déjà dans…

### PR DESCRIPTION
**Pull request by lpereir4**
… le revue d'importation

### CODE REVIEW

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by lpereir4** [comment]
@jdauphant Tu peux faire une review stp?

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

File: app/views/reviewUsersImport.scala.html [hunk 1]
```diff
@@ -12,41 +14,105 @@
...
-      </td>
-      <td class="mdl-data-table__cell--non-numeric cell-import-user">
-        @helper.input(user("user")("phone-number"), 'label -> "Numéro de téléphone", 'class -> "input-import-user") { (id, name, value, args) =>
-        <input class="mdl-textfield__input input-import-user" type="text" name="@name" id="@id" value="@value" @toHtmlArgs(args)>
-        }
-      </td>
-      <td class="mdl-data-table__cell--non-numeric cell-import-user">
-        @helper.input(user("user")("quality"), 'label -> "Qualité", 'class -> "input-import-user") { (id, name, value, args) =>
-        <input class="mdl-textfield__input input-import-user" type="text" name="@name" id="@id" value="@value" @toHtmlArgs(args)>
+      @defining(userEmailToAlreadyExistingUser.get(user("user")("email").value.getOrElse(""))) { existingUser =>
```
Comment (jdauphant) [inline app/views/reviewUsersImport.scala.html]: user("user")("email").value.flatMap(userEmailToAlreadyExistingUser.get)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

File: app/views/reviewUsersImport.scala.html [hunk 2]
```diff
@@ -12,41 +14,105 @@
...
+        @if(existingUser.isEmpty) {
```
Comment (jdauphant) [inline app/views/reviewUsersImport.scala.html]: ça ne marche pas sans différencier le formulaire ? (en ajoutant l'id en champs caché et un boolean caché ? vois sans boolean ça doit marcher)

File: app/views/reviewUsersImport.scala.html [hunk 3]
```diff
@@ -12,41 +14,105 @@
...
+                <input class="mdl-textfield__input input-import-user is-disabled mdl-textfield__input--readonly"
+                       type="text"
+                       name="@name"
+                       id="@id"
+                       readonly
+                       value="@existingUser.map(_.name).getOrElse("")" @toHtmlArgs(args)>
+              }
+            </td>
+            <td class="mdl-data-table__cell--non-numeric cell-import-user">
+              <a href="@existingUser.map(user => s"/utilisateurs/${user.id}").getOrElse("")">@existingUser.map(_.email).getOrElse("")</a>
```
Comment (jdauphant) [inline app/views/reviewUsersImport.scala.html]: target="_ blank" pour ouvrir dans une nouvelle fenêtre

File: app/views/reviewUsersImport.scala.html (hunk 2 above)
Comment (lpereir4) [reply app/views/reviewUsersImport.scala.html]: J'ai préféré distinguer les deux cas, ça rend le code plus lisible.

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

File: app/views/reviewUsersImport.scala.html (hunk 2 above)
Comment (jdauphant) [reply app/views/reviewUsersImport.scala.html]: Je suis plutôt de l'avis inverse, "do not repeat yourself". Distinguer les cas ne simplifie pas la maintenance, on peut trouver d'autres combines pour améliorer la visibilité si besoins.
La section est quasi la même, on peut la traité en appliquant une légere différence.

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

File: app/views/reviewUsersImport.scala.html (hunk 2 above)
Comment (lpereir4) [reply app/views/reviewUsersImport.scala.html]: @jdauphant Je viens de condenser les choses.

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

File: app/views/reviewUsersImport.scala.html [hunk 4]
```diff
@@ -72,41 +103,87 @@
...
+                }
+              </div>
+            } else {
+              @defining(existingGroup.map(_.organisation).get) { value =>
+                <div class="mdl-cell mdl-cell--4-col">
+                  Organisation :
+                  <select readonly id="@group("group")("organisation").id" name="@group("group")("organisation").name">
+                    <option style="font-weight: bold" value="">Organisme non-référencé</option>
+                    @for(organisation <- Organisation.all) {
+                      <option value="@organisation.shortName" @if(value.exists(value => organisation.name.contains(value) || organisation.shortName.contains(value))) {selected}>
```
Comment (jdauphant) [inline app/views/reviewUsersImport.scala.html]: L142 : 
``` <option value="@organisation.shortName" @if(field.value.exists(value => organisation.name.contains(value) || organisation.shortName.contains(value))) {selected}>```

-> existingGroup.map(_.organisation.name).orElse(field.value)
ça évite de recopier tous le bloc et de dupliquer la majeur partie du segment.

Et si tu veux vraiment séparer, il faudrait  faire des fonctions avec le framework.
organisationSelect(organisations: List[Organisation], selected: String, readOnly: Boolean = false)

File: app/views/reviewUsersImport.scala.html [hunk 5]
```diff
@@ -72,41 +103,87 @@
...
-                  <option style="font-weight: bold" value="">Organisme non-référencé</option>
-                  @for(organisation <- Organisation.all) {
-                    <option value="@organisation.shortName" @if(field.value.exists(value => organisation.name.contains(value) || organisation.shortName.contains(value))) {selected}>
-                      @organisation.shortName : @organisation.name
-                    </option>
+              <div class="mdl-cell mdl-cell--12-col">
+                Territoires : @for(areaId <- existingGroup.map(_.areaIds).getOrElse(Nil)) {
+                  @for(area <- Area.fromId(areaId)) {
+                    <b>@{area.name} (@area.inseeCode) /</b>
                   }
```
Comment (jdauphant) [inline app/views/reviewUsersImport.scala.html]: Il faudrait évider de duplique le code. Tu peux soit mixer avec des orElse sur les options, faire des fonction twirl et apparemment tu dois pouvoir definir des variables twirl de cette manière :
```
@areaIds = @{existingGroup match {
  case Some(group) => group.areaIds
  case None => helper.repeat(group("group")("area-ids")).map(_.value).flatMap(UUIDHelper.fromString)
}}
```
Source : https://stackoverflow.com/a/41534679

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

File: app/views/reviewUsersImport.scala.html (hunk 5 above)
Comment (lpereir4) [reply app/views/reviewUsersImport.scala.html]: @jdauphant Il s'avère que repeat renvoie une liste de Html et non les éléments du formulaire (https://github.com/playframework/playframework/blob/94c2e20c6cfbb5f3a5ff125adb9a5d9df24a8bcb/core/play/src/main/scala/views/helper/Helpers.scala#L94) je ne peux donc pas factoriser le calcul de la collection de territoire vu qu'il est lié à la génération du html.

Comment (jdauphant) [reply app/views/reviewUsersImport.scala.html]: Ok bon à savoir

**Comment by jdauphant** [comment]
J'ai fait une proposition dans une PR https://github.com/betagouv/aplus/pull/415 , ça finalise si c'est bon on peut tout merger

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 4
           

Clones added
============
- app/services/UserService.scala  1
- app/controllers/UserController.scala  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 4
           

Clones added
============
- app/controllers/UserController.scala  5
- app/services/UserService.scala  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 4
           

Clones added
============
- app/services/UserService.scala  1
- app/controllers/UserController.scala  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by lpereir4** [comment]
> En fait, il y a bug.
> S'il y a une erreur dans le formulaire, les utilisateurs et groupes existant sont considéré comme utilisateurs à ajouter :
> Formulaire :
> image
> Et quand je valide avec une erreur :
> image
> 
> Pour corriger ça, une solution serait d'avoir un boolean "alreadyExist" sur les utilisateurs, les groupes dans le formulaire et de se baser sur ça dans le formulaire (Et limiter, voir ne pas utiliser les "groupNameToAlreadyExistingGroup").
> Je n'en ai pas d'autre en tête.

J'ai modifié le modèle et le mapping pour que le formulaire intègre l'information 'alreadyExist'.

**Comment by jdauphant** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Clones added
============
- app/controllers/UserController.scala  14
- app/services/UserService.scala  1
- app/helper/CsvHelper.scala  5
- app/controllers/CSVImportController.scala  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/aplus/aplus/pullRequest?prid=4756029&bid=15862017)

**Comment by lpereir4** [comment]
> Il faudrait que le flag "alreadyExists" soit directement "UserGroupFormData" et "UserFormData", ça éviterait d'avoir à le mettre en base de donnée (ça devrait simplifier le code).

@jdauphant c'est fait.

**Pull request merged**